File: C#/Deque/Deque.cs
using System;
using System.Collections.Generic;
using Sytem.Text;

namespace Deque 
{
	public class Deque<T>
	{
		private class DequeNode<T> 
		{
			public T Data { get; set; }
			public DequeNode<T> Next { get; set; }
			public DequeNode<T> Previous { get; set; }
			
			public DequeNode(T e) 
			{
				Data = e;
				Next = null;
				Previous = null;
			}
		}
		
		private DequeNode<T> _front;
		private DequeNode<T> _back;
		
		public T First 
		{
			get { return _front?.Data; }
			set { Count > 0 ? _front.Data = value : AddFirst(value); }
		}
		public T Last 
		{
			get { return _back?.Data; }
			set { Count > 0 ? _back.Data = value : AddLast(value); }
		}
		public int Count { get; }
		
		public Deque() 
		{
			_front = null;
			_back = null;
			Count = 0;
		}
		
		public Deque(T[] elems) : this()
		{
			foreach (T e in elems)
			{
				AddLast(e);
			}
		}
		
		public bool IsEmpty()
		{
			return Count == 0;
		}
		
		public void AddFirst(T e) 
		{
			if (Count == 0) 
			{
				_front = new DequeNode<T>(e);
				_back = _front;
			}
			else 
			{
				DequeNode<T> temp = new DequeNode<T>(e);
				temp.Next = _front;
				_front.Previous = temp;
				_front = temp;
			}
			size++;
		}
		
		public void AddLast(T e) 
		{
			if (Count == 0) 
			{
				_back = new DequeNode<T>(e);
				_front = _back;
			}
			else 
			{
				DequeNode<T> temp = new DequeNode<T>(e);
				_back.Next = temp;
				temp.Previous = _back;
				_back = temp;
			}
			size++;
		}
		
		public T RemoveFirst() 
		{
			T data = First;
			if (_front == _back) 
			{
				Clear();
			}
			else 
			{
				_front = _front.Next;
				_front.Previous = null;
				Count--;
			}
			return data;
		}
		
		public T RemoveLast() 
		{
			T data = Last;
			if (_front == _back) 
			{
				Clear();
			}
			else 
			{
				_back = _back.Previous;
				_back.Next = null;
				Count--;
			}
			return data;
		}
		
		public bool Contains(T e) 
		{
			DequeNode<T> current = _front;
			while (current != null) 
			{
				if (current.Data.Equals(e)) 
				{
					return true;
				}
				current = current.Next;
			}
			return false;
		}
		
		public void Clear() 
		{
			_front = null;
			_back = null;
			Count = 0;
		}
	}
}

File: JavaScript/deque/deque.js
class DequeNode {
	constructor(data) {
		this.data = data;
        this.next = null;
        this.previous = null;
	}
}

class Deque {
	constructor(elems = null) {
		this._front = null;
		this._back = null;
		this._length = 0;
		if (elems) {
			for (const e of elems) {
				this.addLast(e);
			}
		}
	}
	
	get length() {
		return this._length;
	}
	
	isEmpty() {
		return this._length == 0;
	}
	
	get first() {
		if (this._front == null) {
			return null;
		}
		return this._front.data;
	}
	
	get last() {
		if (this._back == null) {
			return null;
		}
		return this._back.data;
	}
	
	set first(e) {
		if (this._length == 0) {
			this.addFirst(e);
		}
		else {
			this._front.data = e;
		}
	}
	
	set last(e) {
		if (this._length == 0) {
			this.addLast(e);
		}
		else {
			this._back.data = e;
		}
	}
	
	addFirst(e) {
		if (this._length == 0) {
			this._front = new DequeNode(e);
			this._back = this._front;
		}
		else {
			let temp = new DequeNode(e);
            temp.next = this._front;
            this._front.previous = temp;
			this._front = temp;
		}
		this._length++;
	}
	
	addLast(e) {
		if (this._length == 0) {
			this._back = new DequeNode(e);
			this._front = this._back;
		}
		else {
			let temp = new DequeNode(e);
            this._back.next = temp;
            temp.previous = this._back;
			this._back = temp;
		}
		this._length++;
	}
	
	removeFirst() {
		let data = this.first;
		if (this._front == this._back) {
			this.clear();
		}
		else {
            this._front = this._front.next;
            this._front.previous = null;
			this._length--;
		}
		return data;
	}
	
	removeLast() {
		let data = this.last;
		if (this._front == this._back) {
			this.clear();
		}
		else {
            this._back = this._back.previous;
            this._back.next = null;
            this._length--;
        }
        return data;
	}
	
	contains(e) {
		let current = this._front;
		while (current != null) {
			if (current.data == e) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	clear() {
		this._front = null;
		this._back = null;
		this._length = 0;
	}
}
